schema/crdb/README.adoc

```diff
@@ -8,7 +8,16 @@ This directory describes the schema(s) used by CockroachDB.
 We use the following conventions:
 
 * `schema/crdb/VERSION/up.sql`: The necessary idempotent migrations to transition from the
-  previous version of CockroachDB to this version.
+  previous version of CockroachDB to this version. These migrations will always be placed
+  within one transaction per file.
+** If more than one change is needed per version, any number of files starting with `up`
+   and ending with `.sql` may be used. These files will be sorted in lexicographic order
+   before being executed, and each will be executed in a separate transaction.
+** CockroachDB documentation recommends the following: "Execute schema changes... in a single
+   explicit transaction consisting of the single schema change statement".
+   Practically this means: If you want to change multiple tables, columns,
+   types, indices, or constraints, do so in separate files.
```

Comment:
This seems too strong — you can do a bunch of column changes in a single `alter table`. I wouldn't want people to think they have to do one `drop column` per file. Maybe we can make the rule of thumb one statement per file?

Reply:
Can you find the source for this? I saw that there was an exception if you're creating a table, but where are we clear to do multiple column changes?

Reply:
For example:

https://www.cockroachlabs.com/docs/stable/online-schema-changes#:~:text=DROP%20COLUMN%20can%20result%20in%20data%20loss%20if%20one%20of%20the%20other%20schema%20changes%20in%20the%20transaction%20fails%20or%20is%20canceled.%20To%20work%20around%20this%2C%20move%20the%20DROP%20COLUMN%20statement%20to%20its%20own%20explicit%20transaction%20or%20run%20it%20in%20a%20single%20statement%20outside%20the%20existing%20transaction.

> [DROP COLUMN](https://www.cockroachlabs.com/docs/v23.1/alter-table#drop-column) can result in data loss if one of the other schema changes in the transaction fails or is canceled. To work around this, move the DROP COLUMN statement to its own explicit transaction or run it in a single statement outside the existing transaction.

Reply:
Ah, I see https://www.cockroachlabs.com/docs/v23.1/alter-table#add-and-rename-columns-atomically , but I think that the "DROP COLUMN" case is still an exception, and any modifications to constraints and indices also probably need to be separate?

Reply:
... All this is to say: I think there may be exceptions, but I'd rather folks stick to "one DDL per file" until they have a citation from the CRDB docs telling them it's safe to do something else. This stuff is super hard to get right, and if we have to lean in any direction, I'd rather we be conservative.

Reply:
https://github.com/oxidecomputer/omicron/blob/03b7588bc365a5afa8f0069347dadc80f8636454/schema/crdb/3.0.1/up1.sql

```sql
ALTER TABLE omicron.public.ip_pool
    DROP COLUMN IF EXISTS project_id,
    ADD COLUMN IF NOT EXISTS is_default BOOLEAN NOT NULL DEFAULT FALSE,
    DROP CONSTRAINT IF EXISTS project_implies_silo,
    DROP CONSTRAINT IF EXISTS internal_pools_have_null_silo_and_project;
```

...is that bad?

Reply:
So, for what it's worth, I think a lot of the failures that CRDB docs are nervous about are "more real" in the context of live update, and are **really** bad if you're doing a DDL + adding an invariant (like a unique index, or constraint) concurrently, where you could insert data that violates that constraint before the upgrade finishes.

So when I say "bad", I want to answer from that context. "Bad" may be "not actually that bad" when we're doing our current "offline upgrade" scheme, where traffic is basically halted until a single node finishes doing update.

But:

- As far as I can tell, the risk of the `DROP COLUMN` is that it may partially occur, if one of the other DDL statements fails and otherwise rolls back. This could result in a state where the IP pool `project_id` is... partially dropped?  I don't actually know how this would manifest.
- The ordering of "adding and removing constraints" is kinda fuzzy to me. https://www.cockroachlabs.com/docs/stable/online-schema-changes#schema-change-ddl-statements-inside-a-multi-statement-transaction-can-fail-while-other-statements-succeed talks at length about scenarios where similar looking statements can insert {data, constraint on that data} concurrently, but the evaluation is not done until the end of the transaction, and rollback cannot occur. In your **specific** case, I don't see how it could break, but CRDB  also says "This error will occur in various scenarios, including but not limited to...", so I dunno about all potential errors that **could** occur.


All this is to say, the following **might** be safer, in different files. Is it **required**? I honestly don't know.

```sql
ALTER TABLE omicron.public.ip_pool
    DROP COLUMN IF EXISTS project_id,

ALTER TABLE omicron.public.ip_pool
    ADD COLUMN IF NOT EXISTS is_default BOOLEAN NOT NULL DEFAULT FALSE,

ALTER TABLE omicron.public.ip_pool
    DROP CONSTRAINT IF EXISTS project_implies_silo,

ALTER TABLE omicron.public.ip_pool
    DROP CONSTRAINT IF EXISTS internal_pools_have_null_silo_and_project;
```